src/graft_dag.cpp

```diff
@@ -0,0 +1,487 @@
+  // Assert that at least one of the nodes are not in the host DAG or graft DAG.
+  const bool parent_is_new =
+      !(host_dag_.ContainsNode(parent_subsplit) || ContainsGraftNode(parent_subsplit));
+  const bool child_is_new =
+      !(host_dag_.ContainsNode(child_subsplit) || ContainsGraftNode(child_subsplit));
+  // Assert that parent and child are a valid node pair to add to DAG.
+  if (host_dag_.IsValidAddNodePair(parent_subsplit, child_subsplit) == false) {
+    Failwith("AddGraftNodePair(): Is not valid node pair.");
+  }
+  // Assert that parent and child don't already exist in the DAG.
+  if (!parent_is_new && !child_is_new) {
+    return results;
+  }
+  // Find relationship between nodes.
+  bool is_left = child_subsplit.SubsplitIsLeftChildOf(parent_subsplit);
+  // Add nodes.
+  if (parent_is_new) {
+    CreateGraftNode(parent_subsplit);
+  }
+  const size_t parent_id = GetDAGNodeId(parent_subsplit);
+  if (child_is_new) {
+    CreateGraftNode(child_subsplit);
+  }
+  const size_t child_id = GetDAGNodeId(child_subsplit);
+  // Find all adjacent host nodes and add edges to fully connect to new graft nodes.
+  if (parent_is_new) {
+    ConnectNodeToHost(parent_id, parent_subsplit, child_id);
+  }
+  if (child_is_new) {
+    ConnectNodeToHost(child_id, child_subsplit, parent_id);
+  }
+  // Connect parent to child.
+  CreateGraftEdge(parent_id, child_id, is_left);
+
+  return results;
+}
+
+void GraftDAG::RemoveAllGrafts() {
+  graft_nodes_.clear();
+  graft_edges_.clear();
+  subsplit_to_id_.clear();
+}
+
+void GraftDAG::ConnectNodeToHost(const size_t node_id, const Bitset node_subsplit,
+                                 std::optional<size_t> node_to_ignore) {
+  const auto [left_parents, right_parents] =
+      host_dag_.BuildParentIdVector(node_subsplit);
+  ConnectNodeToAdjacentHostNodes(node_id, left_parents, false, true, node_to_ignore);
+  ConnectNodeToAdjacentHostNodes(node_id, right_parents, false, false, node_to_ignore);
+  const auto [left_children, right_children] =
+      host_dag_.BuildChildIdVector(node_subsplit);
+  ConnectNodeToAdjacentHostNodes(node_id, left_children, true, true, node_to_ignore);
+  ConnectNodeToAdjacentHostNodes(node_id, right_children, true, false, node_to_ignore);
+}
+
+void GraftDAG::ConnectNodeToAdjacentHostNodes(
+    const size_t main_node_id, const SizeVector adjacent_node_ids,
+    const bool is_main_node_parent, const bool is_left,
+    std::optional<size_t> ignored_node_id_opt) {
+  for (const auto adjacent_node_id : adjacent_node_ids) {
+    if (ignored_node_id_opt && (adjacent_node_id == *ignored_node_id_opt)) {
+      continue;
+    }
+    const size_t parent_id = (is_main_node_parent ? main_node_id : adjacent_node_id);
+    const size_t child_id = (is_main_node_parent ? adjacent_node_id : main_node_id);
+    CreateGraftEdge(parent_id, child_id, is_left);
+  }
+}
+
+// ** Build Indexers/Vectors
+
+std::pair<SizeVector, SizeVector> GraftDAG::BuildParentIdVector(
```

Comment:
```suggestion
std::pair<SizeVector, SizeVector> GraftDAG::BuildParentIdVectors(
```

... there are two of them

Reply:
Note below as well.

Reply:
Renamed.